System id improvements (#4389)
ArduPilot/ardupilot_wiki: copter/source/docs/systemid-mode-ctrl-optimization.rst

| Gain Margin                        | 7 dB            | 7 dB          | Larger is better     |
+------------------------------------+-----------------+---------------+----------------------+
| Step Response: Response Time       | 0.02 s          | 0.2s          | Smaller is better    |
+------------------------------------+-----------------+---------------+----------------------+
| Step Response: Overshoot           | 5 %             | 5 %           | Smaller is better    |
+------------------------------------+-----------------+---------------+----------------------+
| Step Response: Stationary Error    | 5 %             | 0.1 %         | Smaller is better    |
+------------------------------------+-----------------+---------------+----------------------+
| Control Variable Max. Overshoot    | 80 %            | 10 %          | Smaller is better    |
+------------------------------------+-----------------+---------------+----------------------+
| Disturbance Rejection              | see graph below |not applied    | Smaller is better    |
+------------------------------------+-----------------+---------------+----------------------+

> The frequency-based disturbance rejection curve in decibel is shown below.

Reviewer: We decided on the requirements values ourselves.
Yes, these requirements are for a specific craft. Other craft need other values.

As a reference, the blue line represents the rejection gain that is achieved with the default parameters of ArduCopter.
The optimized system is required to stay bellow the yellow area to have a better disturbance rejection.

.. image:: ../images/disturbanceRejectionRequirement.png
:name: fig-dist-reject-req

Optimization Results in the Simulation
--------------------------------------

> The next four plots show the bode plot of the closed-loop system the simulation results for the tracking behavior of a 10 degree step, the normalized controller output corresponding to the angle step as well as the disturbance behavior for the roll axis.

Reviewer 2: ```suggestion
The next four plots show the bode plot of the closed-loop system, the simulation results for the tracking behavior of a 10 degree step, the normalized controller output corresponding to the angle step, and the disturbance behavior for the roll axis.
```

> The normalized controller output evaluates whether the optimized behavior can actually be realized by the real system.

Reviewer 2: I would suggest you do a better job at describing how the plot shows that the optimized behavior can realized by a real system.  It's not obvious to me.

It is assumed that the disturbances act as external torques on the airframe.
Therefore, the disturbance behavior is simulated as a step added to the controller output signal.

Bode plot of the closed-loop system:

> .. image:: ../images/rollAxisClosedLoopBode.png

Reviewer 2: could you add to the graph or in the text what the input (target roll attitude?) and the output (actual roll attitude?) would be for the bode plot?

:name: fig-bode-cl-system

Tracking Behaviour:

> .. image:: ../images/rollAxisTrackingSim.png

Reviewer 2: I find it surprising that the autotune produced a response that was so oscillatory.  I don't think that would be acceptable.  @lthall any thoughts why the autotune would have resulted with a tune that was so oscillatory?  I would think they could do better with the autotune feature.

Reviewer 3: In my opinion this is the main problem with this page. The response of autotune is obviously not based on a good autotune OR the system identification is incorrect.
The results of a successful system identification and optimisation should result in tuning values similar to autotune based on your requirements.
A good autotune should have responses similar to your optimised response. (autotune generates step responses and adjusts gains based on overshoot and the first minimum so it is obvious that the responses shown here are inconsistent with autotune)

Author: The simulation results are based on an attitude controller that was simplified quite a bit since I didn't have time left at the end of my master thesis to realize a proper reconstruction of the controller structure. As an example, the acceleration limits were not implemented in order to linearize the controller for the model-based design. The improvement of the controller used for the simulation will be done in the upcoming work, so we can achieve more reliable simulation results. Additionally, we can provide plots that compare the real tracking behaviour of both controllers.

@lthall I don't fully agree that the requirements for the model-based tune should necessarily result in a similar tune. To my understanding, autotune is evaluating the resulting tune based on the tracking behaviour alone. Contrary to this, the model-based approach considers stability and disturbance behaviour as well. Since these are conflicting design targets, the resulting parameters are likely to differ from the autotune approach. Also, we allow different values for the P and I part of the controller, so the optimization has more freedom in finding good parameter values based on the given requirements.

Reviewer 3: I said similar, not identical. The response should be very close too. You are using the same overshoot requirements and a fast step response time requirement. These two goals will pull the result towards the autotuned result. Phase and gain margin may be a little different to the autotune result but I am guessing that it will be similar (this is one of the things I am interested in seeing how they vary). Again, disturbance rejection isn't going to be pushing anything to far away from the autotuned result.

Step response stationary error should be zero unless I don't understand what that is. I am not confident I do understand what this is. (all errors should decay to zero)
I don't know what "Control Variable Max. Overshoot" is.

The only significant difference I can see between the two will be the different requirement for the I term. Again, this is one of the things I am interested to see how they differ.

+----------------------------------------+-----------+-----------+----------+-------------------+

The following two tables show the optimized flight controller parameters for the roll and yaw axis compared to the default parameters of ArduCopter.
The defaults are also used as initial parameters for the optimization.
As a reference, parameters obtained from an autotune flight with :ref:`AUTOTUNE_AGGR<AUTOTUNE_AGGR>` = 0.1 are also shown.
It shows that the optimized parameters differ greatly from the default and autotune parameters.
Again remember that the presented results are **only valid for the examined quadcopter** with a weight of 1.5 kg, an arm length of 22 cm and 9 inch propellers.
The resulting parameters are specifically applicable for this copter configuration and **must not** be used for any other copter.

Roll:

+-------------------------------------------+-----------------------+----------------------+------------------------+
| Parameter                                 | Default               | Optimized            | Autotune               |
+-------------------------------------------+-----------------------+----------------------+------------------------+

> | :ref:`ATC_RAT_RLL_P<ATC_RAT_RLL_P>`       | 0.1350                | 0.123                | 0.240025               |

Reviewer 4: can I see a log of it flying with I = 8x P ?

Reviewer 2: I believe this is caused by the high disturbance rejection criteria.   I would think this could have some implications with integrator management during sloped takeoffs and landings.

+-------------------------------------------+-----------------------+----------------------+------------------------+
| :ref:`ATC_RAT_RLL_I<ATC_RAT_RLL_I>`       | 0.1350                | 1.0                  | 0.240025               |
+-------------------------------------------+-----------------------+----------------------+------------------------+
| :ref:`ATC_RAT_RLL_D<ATC_RAT_RLL_D>`       | 0.0036                | 0.0053               | 0.007948               |
+-------------------------------------------+-----------------------+----------------------+------------------------+
| :ref:`ATC_RAT_RLL_FF<ATC_RAT_RLL_FF>`     | 0                     | 0                    | 0                      |
+-------------------------------------------+-----------------------+----------------------+------------------------+

> | :ref:`ATC_RAT_RLL_FLTT<ATC_RAT_RLL_FLTT>` | 23.0000               | 50.256               | 5.0                    |

Reviewer 2: How did FLTT get changed in autotune.  As far as I can tell from the code, this is not a filter that is changed during multicopter autotune.  Is this a typo?

Reviewer 5: It's not changed in autotune, it is switched off during twitches however

Reviewer 3: I don't believe this is true. Autotune is run in the linear region of the attitude controller. It has step size limits that attempt to ensure this is the case. The step test does not apply any acceleration limits or other input shaping / control model. In operation the acceleration limits are there to ensure the aircraft is safe when it is recovering from large angle errors. In normal operation and mauvers the acceleration limit in the attitude controller should not be reached.

It looks like you had to reduce the FLTT to 5 Hz to get the autotune results to be stable. That seems very strange to me.

It is possible that your system ID settings may have pushed the aircraft into an acceleration limited region and that has resulted in an inaccurate model.  However, if the model is based on the rate output -> gyro measurement then any controller limits or command model should not impact it. Maybe if the outputs are saturating, that might mess up the system ID result.

The key point is Autotune is designed to optimise the aircraft in the linear region, just as would expect your system ID model to represent. If your model is not stable with the autotuned results, with the same filter settings, then the model is not representative of the aircraft.
